com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -0,0 +1,343 @@
+                                            problemMarkers = sourceFile.findMarkers(
+                                                    IJavaModelMarker.JAVA_MODEL_PROBLEM_MARKER, true,
+                                                    IResource.DEPTH_INFINITE);
+                                            for (IMarker problemMarker : problemMarkers) {
+                                                if (problemMarker.getAttribute(IMarker.SEVERITY,
+                                                        -1) == IMarker.SEVERITY_ERROR) {
+                                                    hasBlockingErrors = true;
+                                                    break;
+                                                }
+                                            }
+                                        }
+                                    } catch (CoreException e) {
+                                        logger.log(Level.SEVERE, "Failed to visit classes: " + e.getMessage(), e);
+                                    }
+                                    if (!hasBlockingErrors) {
+                                        changedFiles.add(resource);
+                                        // dot-delimit the name
+                                        fullyQualifiedNames.add(qualifiedName.replace('/', '.'));
+                                    }
+                                }
+                            }
+                        }
+                        return false;
+
+                    default:
+                        return true;
+                }
+            }
+            return true;
+        }
+
+        /**
+         * Resets the file collection to empty.
+         */
+        public void reset() {
+            changedFiles = new ArrayList<>();
+            fullyQualifiedNames = new ArrayList<>();
+        }
+
+        /**
+         * Answers a collection of changed class files or <code>null</code>.
+         */
+        public List<IResource> getChangedClassFiles() {
+            return changedFiles;
+        }
+
+        /**
+         * Returns a collection of qualified type names corresponding to the changed
+         * class files.
+         *
+         * @return List
+         */
+        public List<String> getQualifiedNamesList() {
+            return fullyQualifiedNames;
+        }
+
+        /**
+         * Returns the source file associated with the given type, or <code>null</code>
+         * if no source file could be found.
+         *
+         * @param project
+         *            the java project containing the classfile
+         * @param qualifiedName
+         *            fully qualified name of the type, slash delimited
+         * @param sourceAttribute
+         *            debug source attribute, or <code>null</code> if none
+         */
+        private IResource getSourceFile(IJavaProject project, String qualifiedName, String sourceAttribute) {
```

Comment:
Looking at the underlying implementation `JavaProject#findElement(IPath)`, it will compute the fully qualified name from path and then using fully qualified name to findType.

So here you can use `JavaProject#findType(fullyQualifiedName)` directly.

Reply:
This is the JDT's code. Leave it as-is since it is fully tested.

Reply:
I don't think JDT's code is right, getSourceFile and findElement in last line are unnecessary. Using findType can get the result directly.

If you keep the original implementation, suggest to put it in additional utility class so that we can recognize what's our own logic.

Reply:
This file is JDT implementation. See the license header. We only add glue code in this file.

Reply:
Since the JDT code is well tested, I'll trust their implementation here. There might be some corner case if JavaProject#findType fail  that we don't know. Nothing is guarantee until we do fully verification on the code.